pkg/appstate/wait.go

```diff
@@ -0,0 +1,228 @@
+				return nil
+			}
+		}
+
+		time.Sleep(WaitForResourceInterval)
+	}
+}
+
+func WaitForServiceToBeReady(clientset kubernetes.Interface, namespace, name string) error {
+	for {
+		r, err := clientset.CoreV1().Services(namespace).Get(context.TODO(), name, metav1.GetOptions{})
+		if err != nil && !kuberneteserrors.IsNotFound(err) {
+			return errors.Wrap(err, "failed to get existing service")
+		}
+
+		if !kuberneteserrors.IsNotFound(err) {
+			state := CalculateServiceState(clientset, r)
+			if state == types.StateReady {
+				return nil
+			}
+		}
+
+		time.Sleep(WaitForResourceInterval)
+	}
+}
+
+func WaitForStatefulSetToBeReady(clientset kubernetes.Interface, namespace, name string) error {
+	for {
+		r, err := clientset.AppsV1().StatefulSets(namespace).Get(context.TODO(), name, metav1.GetOptions{})
+		if err != nil && !kuberneteserrors.IsNotFound(err) {
+			return errors.Wrap(err, "failed to get existing statefulset")
+		}
+
+		if !kuberneteserrors.IsNotFound(err) {
+			state := CalculateStatefulSetState(clientset, namespace, r)
+			if state == types.StateReady {
+				return nil
+			}
+		}
+
+		time.Sleep(WaitForResourceInterval)
+	}
+}
+
+func WaitForGenericResourceToBeReady(dr dynamic.ResourceInterface, name string) (err error) {
+	for {
+		_, err := dr.Get(context.TODO(), name, metav1.GetOptions{})
+		if err != nil && !kuberneteserrors.IsNotFound(err) {
+			return errors.Wrap(err, "failed to get existing resource")
+		}
+
+		if !kuberneteserrors.IsNotFound(err) {
+			return nil
+		}
+
+		time.Sleep(WaitForResourceInterval)
+	}
+}
+
+func WaitForProperty(namespace, name string, gvk *schema.GroupVersionKind, path, desiredValue string) error {
+	dr, err := k8sutil.GetDynamicResourceInterface(gvk, namespace)
+	if err != nil {
+		return errors.Wrap(err, "failed to get dynamic resource interface")
+	}
+
+	for {
+		r, err := dr.Get(context.TODO(), name, metav1.GetOptions{})
+		if err != nil && !kuberneteserrors.IsNotFound(err) {
+			return errors.Wrap(err, "failed to get existing resource")
+		}
+
+		if !kuberneteserrors.IsNotFound(err) {
+			if matches, err := resourcePropertyMatchesValue(r, path, desiredValue); err != nil {
+				return errors.Wrap(err, "failed to check if resource property matches value")
+			} else if matches {
+				return nil
+			}
+		}
+
+		time.Sleep(WaitForResourceInterval)
+	}
+}
+
+func resourcePropertyMatchesValue(r *unstructured.Unstructured, path, desiredValue string) (bool, error) {
+	if path == "" {
+		return false, errors.New("key cannot be empty")
+	}
+
+	parser := jsonpath.New("wait-for-property")
+	if err := parser.Parse(fmt.Sprintf("{ %s }", path)); err != nil {
+		return false, errors.Wrapf(err, "failed to parse jsonpath %s", path)
+	}
+
+	buf := new(bytes.Buffer)
+	err := parser.Execute(buf, r.Object)
+	if err != nil {
+		// don't return an error here since the field may not exist yet
+		logger.Warnf("failed to execute jsonpath: %v", err)
+	}
+
+	if buf.String() == desiredValue {
+		return true, nil
+	}
```

Comment:
just equality might be limiting. I can see other use cases like `>= 0` 

have you considered using [CEL](https://github.com/google/cel-spec)? It's being adopted by the k8s project for native resources, so it should be a safe bet. It would also make sense in other places throughout the codebase, such as analyzers (for preflight checks)

Reply:
thanks for sharing this.  likely something we'll look at for a future iteration on this feature